core/commands/dag/dag.go

```diff
@@ -668,3 +671,134 @@ The output of blocks happens in strict DAG-traversal, first-seen, order.
 		},
 	},
 }
+
+type DagStat struct {
+	Size      uint64
+	NumBlocks int64
+}
+
+func (s *DagStat) String() string {
+	return fmt.Sprintf("Size: %d, NumBlocks: %d", s.Size, s.NumBlocks)
+}
+
+var DagStatCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Gets stats for a DAG",
+		ShortDescription: `
+'ipfs dag size' fetches a dag and returns various statistics about the DAG.
+Statistics include size and number of blocks.
+
+Note: This command skips duplicate blocks in reporting both size and the number of blocks
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("root", true, false, "CID of a DAG root to get statistics for").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Return progressive data while reading through the DAG").WithDefault("true"),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		progressive := req.Options[progressOptionName].(bool)
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		rp, err := api.ResolvePath(req.Context, path.New(req.Arguments[0]))
+		if err != nil {
+			return err
+		}
+
+		if len(rp.Remainder()) > 0 {
+			return fmt.Errorf("cannot return size for anything other than a DAG with a root CID")
+		}
+
+		obj, err := api.Dag().Get(req.Context, rp.Cid())
+		if err != nil {
+			return err
+		}
+
+		dagstats := &DagStat{}
+		err = traverse.Traverse(obj, traverse.Options{
+			DAG:   api.Dag(),
+			Order: traverse.DFSPre,
+			Func: func(current traverse.State) error {
+				stat, err := current.Node.Stat()
+				if err != nil {
+					return err
+				}
+
+				// stat.BlockSize gives the raw size of the data which is what we want here.
+				// However, node.Stat() is not implemented by any node types other than DagPB, which defines Size() as
+				// a field stored in the data instead of the raw size of the data itself.
+				//
+				// Therefore, we check if stat is defined and if so use `stat.BlockSize` and otherwise just rely on the
+				// Size() function.
+				if len(stat.Hash) == 0 {
+					size, err := current.Node.Size()
+					if err != nil {
+						return err
+					}
+					dagstats.Size += size
+				} else {
+					dagstats.Size += uint64(stat.BlockSize)
+				}
```

Comment:
This is dumb. @Stebalien do you know if there's a better/cleaner way to handle this? Should I just be checking if the node is DagPB instead of checking `len(stat.Hash)!=0`?

The latter seemed slightly more generic since it appears that the pattern is either to return a full stat object or an empty stat object with nothing in between.

Reply:
I'd drop the `Stat` call entirely. It's deprecated from IPLD's standpoint.

Reply:
I.e., just get the node size.

Reply:
I can't do `node.Size` for DagPB since it will give me the cumulative node size https://github.com/ipfs/go-merkledag/blob/749fd8717d46b4f34c9ce08253070079c89bc56d/node.go#L236.

Were you thinking it'd be better to do `len(nd.RawData())`?

Reply:
I really hate this code... Yes, `len(nd.RawData())` is the only way to go.